src/app-layer-enip-common.c

```diff
@@ -0,0 +1,904 @@
+            SCLogDebug(
+                    "DecodeCommonPacketFormat - CONNECTED DATA ITEM - parse CIP\n");
+            DecodeCIPPDU(input, input_len, enip_data, offset);
+            break;
+        case UNCONNECTED_DATA_ITEM:
+            SCLogDebug("DecodeCommonPacketFormat - UNCONNECTED DATA ITEM\n");
+            DecodeCIPPDU(input, input_len, enip_data, offset);
+            break;
+        default:
+            SCLogDebug("DecodeCommonPacketFormat - UNKNOWN TYPE 0x%x\n\n",
+                    enip_data->encap_data_item.type);
+            return 0;
+    }
+
+    return 1;
+
+}
+
+
+
+/**
+ * \brief Decode CIP packet
+ * @param input, input_len data stream
+ * @param enip_data stores data from Packet
+ * @param offset current point in the packet
+ * @return 1 Packet ok
+ * @return 0 Packet has errors
+ */
+
+int DecodeCIPPDU(uint8_t *input, uint32_t input_len,
+        ENIPTransaction *enip_data, uint16_t offset)
+{
+    int ret = 1;
+
+    if (enip_data->encap_data_item.length == 0)
+    {
+        SCLogDebug("DecodeCIP: No CIP Data\n");
+        return 0;
+    }
+
+    if (offset > (input_len - sizeof(uint8_t)))
+    {
+        SCLogDebug("DecodeCIP: Parsing beyond payload length\n");
+        return 0;
+    }
+
+    uint8_t service = 0;
+    service = *(input + offset);
+
+    //SCLogDebug("CIP Service 0x%x\n", service);
+
+    //use service code first bit to determine request/response, no need to save or push offset
+    if (service >> 7)
+    {
+        ret = DecodeCIPResponsePDU(input, input_len, enip_data, offset);
+    } else
+    {
+        ret = DecodeCIPRequestPDU(input, input_len, enip_data, offset);
+    }
+
+    return ret;
+}
+
+
+
+/**
+ * \brief Decode CIP Request
+ * @param input, input_len data stream
+ * @param enip_data stores data from Packet
+ * @param offset current point in the packet
+ * @return 1 Packet ok
+ * @return 0 Packet has errors
+ */
+int DecodeCIPRequestPDU(uint8_t *input, uint32_t input_len,
+        ENIPTransaction *enip_data, uint16_t offset)
+{
+    int ret = 1;
+
+    if (enip_data->encap_data_item.length < sizeof(CIPReqHdr))
+    {
+        SCLogDebug("DecodeCIPRequest - Malformed CIP Data\n");
+        return 0;
+    }
+
+    uint8_t service; //<-----CIP SERVICE
+    uint8_t path_size;
+    if (offset >= (input_len - sizeof(uint8_t)*2))
+    {
+        SCLogDebug("DecodeCIPRequest: Parsing beyond payload length\n");
+        return 0;
+    }
+    ENIPExtractUint8(&service, input, &offset);
+    ENIPExtractUint8(&path_size, input, &offset);
+
+    if (service > MAX_CIP_SERVICE)
+    { // service codes of value 0x80 or greater are not permitted because in the CIP protocol the highest order bit is used to flag request(0)/response(1)
+        SCLogDebug("DecodeCIPRequest - INVALID CIP SERVICE 0x%x\n", service);
+        return 0;
+    }
+
+    //reached maximum number of services
+    if (enip_data->service_count > 32)
+    {
+        SCLogDebug("DecodeCIPRequest: Maximum services reached\n");
```

Comment:
so far we haven't seen any ENIP packets so far with more than 32 services so we decided to put a limit or it could possibly go on forever.